docs/enriching-your-data/available-enrichments/custom-javascript-enrichment/index.md

```diff
@@ -4,260 +4,28 @@ date: "2020-02-14"
-
-JavaScript enrichment uses [Nashorn Engine](https://docs.oracle.com/javase/10/nashorn/introduction.htm) and since version 3.0.0 of Enrich, many features of ECMAScript 6 are supported. Please refer to [this page](http://openjdk.java.net/jeps/292) to know what is supported. About the uncertainty mentioned, our testing shows that classes and generators don't work, but tail calls do.
-
-:::
+With this enrichment, you can write a JavaScript function to be executed for each event. Use this enrichment to apply your own business logic to your events, including:
+* [Adding extra data](/docs/enriching-your-data/available-enrichments/custom-javascript-enrichment/writing/index.md#adding-extra-entities-to-the-event) to the event in the form of [entities](/docs/understanding-tracking-design/understanding-events-entities/index.md)
+* [Modifying event fields](/docs/enriching-your-data/available-enrichments/custom-javascript-enrichment/writing/index.md#modifying-event-fields-directly) directly
+* [Discarding the event](/docs/enriching-your-data/available-enrichments/custom-javascript-enrichment/writing/index.md#discarding-the-event) so that it goes to [failed events](/docs/managing-data-quality/failed-events/understanding-failed-events/index.md) rather than your data destination
 
 ## Configuration
 
 - [schema](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow/javascript_script_config/jsonschema/1-0-0)
 - [example](https://github.com/snowplow/enrich/blob/master/config/enrichments/javascript_script_enrichment.json)
 
-## Simple Example
-
-Although the use cases for this enrichment can be endless, they don’t need to only serve analytics purposes. A very useful case might be in anti-spoofing for instance. You might want to make sure that the events being sent over are from your app or website rather than one spoofed by a hacker.
-
-To do so you might write a Javascript function that checks if a secret key is valid server side, and only then set the event with the correct “app_id” variable that denotes it is a valid event from your property.
-
-Taking this example further, you could do something like notify the user that sent the event that they are likely on a spoofing website.
-
-Here’s an example snippet of code relating to the above:
-
-```javascript
-var SECRET_APP_ID = "Joshua";
-    
-/**
- * Performs two roles:
- * 1. If this is a server-side event, we
- *    validate that the app_id is our
- *    valid secret. Prevents spoofing of
- *    our server-side events
- * 2. If app_id is not null, return a new
- *    Acme context, derived_app_id, which
- *    contains the upper-cased app_id
- */
- 
-function process(event) {
-    var appId = event.getApp_id();
-  
-    if (platform == "server" && appId != SECRET_APP_ID) {
-        throw "Server-side event has invalid app_id: " + appId;
-    }
-  
-    if (appId == null) {
-        return [];
-    }
-  
-    var appIdUpper = new String(appId.toUpperCase());
```

Comment:
TODO: Does anyone know why we need `new String`?

Reply:
I am assuming https://docs.oracle.com/javase/8/docs/technotes/guides/scripting/nashorn/api.html#sthref21?

Reply:
Can’t reproduce... maybe only relevant for older versions of Nashorn?

Reply:
@stanch, if I'm not mistaken you have to use `new String` for JSON fields `contexts` and `unstruct_event`. For example, `event.setContexts(new String(JSON.stringify(jsonContexts)));`. It won't hurt if you use it also for the rest - POJO fields, `event.setUser_ipaddress(new String(some_string));`. 

In your example, both should work.

Reply:
I tried `setContexts` without `new String` and it worked just fine... Do you know the reason why it’s necessary?

Reply:
In my mind, any member of the enriched TSV is an object. Hence `new String` instead of just string literal. While the result is the same the semantic is different. In particular, a string literal could be [interned](http://en.wikipedia.org/wiki/String_interning) while every field out of 131 TSV fields is an independent entity.

Having said that, I would check with the developers for the exact reason. Mind you, @miike is planning to write [playbooks on enrichments and other components of the pipeline](https://snowplow.slack.com/archives/CDC2DST35/p1680128597700379). You might want to join the forces.

Reply:
Thanks @ihortom!

> In my mind, any member of the enriched TSV is an object. Hence new String instead of just string literal. While the result is the same the semantic is different. In particular, a string literal could be [interned](http://en.wikipedia.org/wiki/String_interning) while every field out of 131 TSV fields is an independent entity.

I am not convinced by this argument:
* All strings in Java _are_ objects, including string literals, interned or not.
* There are no string literals in the examples in question.
* Strings in Java are immutable, so I don’t see a reason why it should matter if a string is interned or not. (Immutability is why strings can be interned in the first place.) If anything, interned strings save on memory.
* If semantics of interning were important, Java developers would be defensively wrapping their strings in `new String`, but I don’t think that happens.

I found a link to [this post](https://web.archive.org/web/20171128004708/http://nelsonwells.net/2012/02/json-stringify-with-mapped-variables/) in one of the JavaScript enrichment code samples. However, it’s from 2012 and uses Rhino and not Nashorn.

I have tested several supposedly problematic scenarios and found all of them to work without `new String`:
* `event.getApp_id().toUpperCase()` (the example in the old docs)
* `event.setContexts(JSON.stringify(...))` (your suggestion)
* `JSON.stringify(event.getApp_id().toUpperCase())` (similar to the linked blog post)

I would not want to recommend this practice without solid reasoning, as it can have negative impact on performance (due to unnecessarily creating new objects). Checking with the team if anyone has more background context on this.

> Mind you, @miike is planning to write [playbooks on enrichments and other components of the pipeline](https://snowplow.slack.com/archives/CDC2DST35/p1680128597700379). You might want to join the forces.

Will definitely sync with Mike. But since you were the one asking for better materials on JS enrichments in that thread, I’d like to put you on the spot and ask what you think of the rest of this PR! :)

Reply:
I guess your initial assumption about the old version would be the reason for using `new String`. I do remember enrichment not working without it.

> since you were the one asking for better materials on JS enrichments in that thread, I’d like to put you on the spot and ask what you think of the rest of this PR! :)

It is the first time I see all the possibilities that JS enrichment can handle documented :). Thus, it is the first time ever I am happy with our guide. Thank you for doing this.